src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Extensions.Internal;
+
+namespace System.Net
+{
+    [EventSource(Name = "System.Net.NameResolution")]
+    internal sealed class NameResolutionTelemetry : EventSource
+    {
+        public static readonly NameResolutionTelemetry Log = new NameResolutionTelemetry();
+
+        public static new bool IsEnabled => Log.IsEnabled();
+
+        private const int ResolutionStartEventId = 1;
+        private const int ResolutionSuccessEventId = 2;
+        private const int ResolutionFailureEventId = 3;
+
+        private PollingCounter? _lookupsRequestedCounter;
+        private EventCounter? _lookupsDuration;
+
+        private long _lookupsRequested;
+
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            if (command.Command == EventCommand.Enable)
+            {
+                // The cumulative number of name resolution requests started since the process started.
+                _lookupsRequestedCounter ??= new PollingCounter("dns-lookups-requested", this, () => Interlocked.Read(ref _lookupsRequested))
+                {
+                    DisplayName = "DNS Lookups Requested"
+                };
+
+                _lookupsDuration ??= new EventCounter("dns-lookups-duration", this)
+                {
+                    DisplayName = "Average DNS Lookup Duration",
+                    DisplayUnits = "ms"
+                };
+            }
+        }
+
+        private const int MaxIPFormattedLength = 128;
+
+        [Event(ResolutionStartEventId, Level = EventLevel.Informational)]
+        public ValueStopwatch ResolutionStart(string hostNameOrAddress)
+        {
+            Debug.Assert(hostNameOrAddress != null);
+
+            if (IsEnabled())
```

Comment:
Don't we need to check that this event level is actually enabled, e.g. `if (IsEnabled(EventLevel.Informational))` in order to avoid the WriteEvent overhead when it's not?

And once we do that, don't we need to separate the incrementing of the counters from the actual event tracing logic?

cc: @brianrob

Reply:
Changed. They are now in the form of
```c#
if (IsEnabled())
{
    Interlocked.Increment(ref _lookupsRequested);

    if (IsEnabled(EventLevel.Informational, EventKeywords.None))
    {
        ResolutionStarting(hostNameOrAddress);
        return ValueStopwatch.StartNew();
    }
}

return default;
```